### PR TITLE
Allow registering multiple standbys concurrently

### LINKED ISSUE #392: Adding a standby while another is in maintenance is not possible and cannot be killed by ctrl+C
Setup:
```bash
pg_autoctl create postgres --pgdata node1 --run
pg_autoctl create postgres --pgdata node2 --run
pg_autoctl enable maintenance --pgdata node2 
```

Then this command fails continously and cannot be killed by ctrl+c anymore

```
pg_autoctl create postgres --pgdata node3
```
Output:
```
^C15:08:02 500 INFO  pg_autoctl received signal SIGINT, terminating
15:08:02 500 ERROR Failed to send signal Interrupt to service postgres with pid 505
15:08:02 506 ERROR Monitor ERROR:  primary node 172.28.149.103:5501 is already in state wait_primary
15:08:02 506 ERROR Monitor HINT:  Retry registering in a moment
15:08:02 506 ERROR SQL query: SELECT * FROM pgautofailover.register_node($1, $2, $3, $4, $5, $6, $7, $8::pgautofailover.replication_state, $9, $10, $11)
15:08:02 506 ERROR SQL params: 'default', '172.28.149.103', '5503', 'postgres', '', '6868238261832629440', '-1', 'init', 'standalone', '50', 'true'
15:08:02 506 WARN  Failed to register node 172.28.149.103:5503 in group -1 of formation "default" with initial state "init" because the monitor is already registering another standby, retrying in 1s
```

Current cluster state
```

  Name |  Node |           Host:Port |       LSN | Reachable |       Current State |      Assigned State
-------+-------+---------------------+-----------+-----------+---------------------+--------------------
node_1 |     1 | 172.28.149.103:5501 | 0/3000728 |       yes |        wait_primary |        wait_primary
node_2 |     2 | 172.28.149.103:5502 | 0/3000728 |       yes |         maintenance |         maintenance
```

### PR DESCRIPTION
**Pull request by DimCitus**
A bug has been found where if we have a single standby and it's in maintenance, we can't add another standby. That's not nice, and to fix it, we need to allow registering a new standby node while the primary is in the WAIT_PRIMARY state. The reason for this state could either be that a standby is in maintenance, or that another standby is concurrently registering to the primary.

To fix, we now lift the limitation that only one standby can be registered at a single time. This way we can register a standby even when the primary is in the WAIT_PRIMARY state already, or even in the JOIN_PRIMARY or APPLY_SETTINGS state. For this to work, some responsibilities are added to the client side of things, where it's now necessary to check for the replication slot creation before moving on, even when the local node has been cleared with registration and assigned CATCHING-UP already.

Fixes #392.

### CODE REVIEW

File: src/bin/pg_autoctl/primary_standby.c [hunk 1]
```diff
@@ -766,8 +809,40 @@ standby_init_database(LocalPostgresServer *postgres,
 	}
 	else
 	{
-		if (!pg_basebackup(pgSetup->pgdata, pgSetup->pg_ctl, upstream))
+		/*
+		 * pg_basebackup has this bug where it will copy over the whole PGDATA
```
Comment (onderkalaci) [inline src/bin/pg_autoctl/primary_standby.c:813]: is there a link to the bug mentioned here? Might be good to add to the comment as well

**Review by onderkalaci — COMMENTED** [review]
I think this looks good overall, though want to think a little more before approving. Sharing some minor notes

**Comment by onderkalaci** [comment]
And, tested this several times with long/short catching up periods & joining at the same time/in different times etc, all worked pretty good

File: src/bin/pg_autoctl/keeper.c [hunk 2]
```diff
@@ -952,20 +942,22 @@ keeper_ensure_configuration(Keeper *keeper, bool postgresNotRunningIsOk)
...
+ * keeper_create_and_drop_replication_slots drops replication slots that we
+ * have on the local Postgres instance when the node is not registered on the
+ * monitor anymore (after a pgautofailover.remove_node() has been issued, maybe
+ * with the command `pg_autoctl drop node [ --destroy ]`); and creates
+ * replication slots for nodes that have been recently registered on the
+ * monitor.
  */
 bool
-keeper_drop_replication_slots_for_removed_nodes(Keeper *keeper)
+keeper_create_and_drop_replication_slots(Keeper *keeper)
```
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/bin/pg_autoctl/keeper.c:953]: Good catch, done.

Side note, that I can't resist sharing: use `grep -c` to count matching lines. Also I use `M-x ack` that uses the [ack](https://manpages.debian.org/buster/ack/ack.1p.en.html) tooling/wrapper, which knows how to handle many source code things, including recursive searches and avoid files you usually don't want to target. Also I guess I could just use `git grep` nowadays, and that offers `git grep -c`.

File: src/bin/pg_autoctl/primary_standby.c [hunk 3]
```diff
@@ -403,19 +445,20 @@ primary_drop_replication_slot(LocalPostgresServer *postgres,
...
 bool
-postgres_replication_slot_drop_removed(LocalPostgresServer *postgres,
-									   NodeAddressArray *nodeArray)
+postgres_replication_slot_create_and_drop(LocalPostgresServer *postgres,
+										  NodeAddressArray *nodeArray)
 {
 	bool result = false;
 	PGSQL *pgsql = &(postgres->sqlClient);
 
 	log_trace("postgres_replication_slot_drop_removed");
```
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/bin/pg_autoctl/primary_standby.c:463]: Fixed.

File: src/bin/pg_autoctl/pgsql.c [hunk 4]
```diff
@@ -1091,6 +1106,42 @@ typedef struct ReplicationSlotMaintainContext
...
+ * slotName exists on the Postgres server.
+ */
+bool
+pgsql_replication_slot_exists(PGSQL *pgsql, const char *slotName,
+							  bool *slotExists)
+{
+	SingleValueResultContext context = { { 0 }, PGSQL_RESULT_BOOL, false };
+	char *sql = "SELECT 1 FROM pg_replication_slots WHERE slot_name = $1";
+	int paramCount = 1;
+	Oid paramTypes[1] = { TEXTOID };
```
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/bin/pg_autoctl/pgsql.c]: Yeah we could, so I've done so.

File: src/bin/pg_autoctl/primary_standby.c [hunk 5]
```diff
@@ -361,6 +361,48 @@ primary_has_replica(LocalPostgresServer *postgres, char *userName, bool *hasStan
 }
 
 
+/*
+ * upstream_has_replication_slot checks whether the upstream server already has
```
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/bin/pg_autoctl/primary_standby.c:365]: Yeah, we could connect to an upstream node that is a replica / standby, such as when in `fast_foward` state. At the moment we don't `pg_basebackup` from a standby, but also, that's definitely something we could envision at some point.

File: src/bin/pg_autoctl/primary_standby.c [hunk 6]
```diff
@@ -361,6 +361,48 @@ primary_has_replica(LocalPostgresServer *postgres, char *userName, bool *hasStan
...
+	char connectionString[MAXCONNINFO] = { 0 };
+
+	/* prepare a PostgresSetup that allows preparing a connection string */
+	strlcpy(upstreamSetup.username, PG_AUTOCTL_REPLICA_USERNAME, NAMEDATALEN);
+	strlcpy(upstreamSetup.dbname, pgSetup->dbname, NAMEDATALEN);
+	strlcpy(upstreamSetup.pghost, primaryNode->host, _POSIX_HOST_NAME_MAX);
+	upstreamSetup.pgport = primaryNode->port;
+	upstreamSetup.ssl = pgSetup->ssl;
+
+	pg_setup_get_local_connection_string(&upstreamSetup, connectionString);
```
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/bin/pg_autoctl/primary_standby.c:390]: Added a comment to make it clearer. Basically re-using code in a context that it was not intended for, but where it's all fine really:

``` c
	/*
	 * Build the connection string as if to a local node, but we tweaked the
	 * pgsetup to target the primary node by changing its pghost and pgport.
	 */
```

File: src/bin/pg_autoctl/primary_standby.c (hunk 1 above)
Comment (DimCitus) [reply src/bin/pg_autoctl/primary_standby.c:817]: No bug report / link that I know of, no, sorry. I intended to work on the fix myself, and then had a chat with folks on the Postgres IRC channel and gave the project away, but I didn't even track if it's been fixed or not yet. It's a very easy fix that's waiting to happen.

Basically `pg_basebackup` calls `waitpid` on the wal receiver child process at https://github.com/postgres/postgres/blob/master/src/bin/pg_basebackup/pg_basebackup.c#L2106 but that's when the file transfer is done already. So if the child process does an early exit because the replication slot does not exists, you see the error message at the top of the logs from `pg_basebackup`, but it still goes on with the base backup, and then fails after having copied a whole PGDATA that it doesn't even know how to re-use on the next attempt.

File: src/monitor/node_active_protocol.c [hunk 7]
```diff
@@ -554,21 +554,6 @@ JoinAutoFailoverFormation(AutoFailoverFormation *formation,
 
 			initialState = REPLICATION_STATE_WAIT_STANDBY;
 
-			/*
```
Comment (onderkalaci) [inline src/monitor/node_active_protocol.c:557]: The comment made me think on the PR again. 
We have [this](https://github.com/citusdata/pg_auto_failover/blob/master/src/monitor/group_state_machine.c#L314-L338) code block in group_state_machine, which is the place that we mark nodes primary/secondary once A secondary  catches up.  
- If we have the other secondary catching up concurrently, the same block will transition to `primary->primary`, `catching up -> secondary`, which is good
-  Even if we add a new standby, the primary switch back to  `wait_primary` and things will  work as expected

Well, this scenario is hard to test locally because even if I have 100GB data, it catches up so fast that I cannot add new standby while others are in progress. 

I wanted to write my understanding here in case this pops-up any edge cases on your hear.

**Review by onderkalaci — APPROVED** [review]
I

File: src/monitor/node_active_protocol.c (hunk 7 above)
Comment (onderkalaci) [reply src/monitor/node_active_protocol.c:557]: > If we have the other secondary catching up concurrently, the same block will transition to primary->primary, catching up -> secondary, which is good

This  is slightly confusing from the user perspective though. Because now we could have `primary` and `catching up` at the same  time

Comment (DimCitus) [reply src/monitor/node_active_protocol.c:557]: Have a look at https://github.com/citusdata/pg_auto_failover/pull/398 where we're creating a tmux script that automates creating a formation. If you remove the `sleep` calls in there, you should be able to generate this race condition more easily.

Otherwise, we have added PRIMARY in the list of states you're showing in that code block in the multiple standby PR, because now we can have standby showing up when the primary is already back to being a primary (failover, report_lsn, back to normal, and a standby was not connecting for reasons and is back just now, would be a scenario, IIRC). So I think we're good here, in particular thanks to the block of code you have highlighted, yes.
